pkg/storage/action_cache_asset_store.go

```diff
@@ -0,0 +1,280 @@
+package storage
+
+import (
+	"context"
+	"time"
+
+	remoteexecution "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/buildbarn/bb-remote-asset/pkg/proto/asset"
+	"github.com/buildbarn/bb-storage/pkg/blobstore"
+	"github.com/buildbarn/bb-storage/pkg/blobstore/buffer"
+	"github.com/buildbarn/bb-storage/pkg/digest"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type actionCacheAssetStore struct {
+	actionCache               blobstore.BlobAccess
+	contentAddressableStorage blobstore.BlobAccess
+	maximumMessageSizeBytes   int
+}
+
+// NewActionCacheAssetStore creates a new AssetStore which stores it's
+// references as ActionResults in the Action Cache.
+func NewActionCacheAssetStore(actionCache, contentAddressableStorage blobstore.BlobAccess, maximumMessageSizeBytes int) AssetStore {
+	return &actionCacheAssetStore{
+		actionCache:               actionCache,
+		contentAddressableStorage: contentAddressableStorage,
+		maximumMessageSizeBytes:   maximumMessageSizeBytes,
+	}
+}
+
+func (rs *actionCacheAssetStore) isDirectory(ctx context.Context, asset *asset.Asset, instance digest.InstanceName) (*remoteexecution.Directory, error) {
+	digest, err := instance.NewDigestFromProto(asset.Digest)
+	if err != nil {
+		return nil, err
+	}
+	directory, err := rs.contentAddressableStorage.Get(ctx, digest).ToProto(&remoteexecution.Directory{}, rs.maximumMessageSizeBytes)
+	if err != nil {
+		return nil, err
+	}
+	return directory.(*remoteexecution.Directory), nil
+}
+
+func (rs *actionCacheAssetStore) actionResultToAsset(ctx context.Context, a *remoteexecution.ActionResult, instance digest.InstanceName) (*asset.Asset, error) {
+	digest := &remoteexecution.Digest{}
+	for _, dir := range a.OutputDirectories {
+		if dir.Path == "out" {
+			digest = dir.TreeDigest
+		}
+	}
+	if digest.Hash != "" {
+		treeDigest, err := instance.NewDigestFromProto(digest)
+		if err != nil {
+			return nil, err
+		}
+		tree, err := rs.contentAddressableStorage.Get(ctx, treeDigest).ToProto(&remoteexecution.Tree{}, rs.maximumMessageSizeBytes)
+		if err != nil {
+			return nil, err
+		}
+		root := tree.(*remoteexecution.Tree).Root
+		digest, err = ProtoToDigest(root)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		for _, file := range a.OutputFiles {
+			if file.Path == "out" {
+				digest = file.Digest
+			}
+		}
+	}
+	return &asset.Asset{
+		Digest:      digest,
+		ExpireAt:    getDefaultTimestamp(),
+		LastUpdated: a.ExecutionMetadata.QueuedTimestamp,
+	}, nil
+}
+
+func (rs *actionCacheAssetStore) Get(ctx context.Context, ref *asset.AssetReference, instance digest.InstanceName) (*asset.Asset, error) {
```

Comment:
It's a little hard to follow what's going on here (and more so in `Put`) due to all the type-changing and error-handling boiler plate. It might improve things to add a couple comments on how it works? In particular the construction of the `AssetReference` file node could do with a quick explanation.